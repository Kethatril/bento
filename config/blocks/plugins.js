const path = require("path"),
    webpack = require("webpack");

const CleanPlugin = require("clean-webpack-plugin");
const CopyGlobsPlugin = require("copy-globs-webpack-plugin");
const WebpackAssetsManifest = require("webpack-assets-manifest");

module.exports = function(config, options, { format }) {
    config.merge({
        plugins: [
            new CleanPlugin(
                [`${path.join(options.get("paths.output"), "**/*")}`],
                {
                    root: options.get("paths.root"),
                    verbose: false
                }
            ),

            new CopyGlobsPlugin({
                pattern: options.get("files.copy"),
                output: `[path]${options.get("filename")}.[ext]`,
                manifest: options.get("manifest")
            }),

            new WebpackAssetsManifest({
                output: "assets.json",
                space: 4,
                writeToDisk: false,
                sortManifest: true,
                assets: options.get("manifest"),
                replacer: format
            }),

            new webpack.DefinePlugin({
                "process.env": {
                    NODE_ENV: JSON.stringify(process.env.NODE_ENV)
                }
            }),

            new webpack.NamedModulesPlugin()
        ]
    });

    if (options.get("env.isProduction")) {
        const OptimizeCssAssetsPlugin = require("optimize-css-assets-webpack-plugin");

        config.merge({
            plugins: [
                new webpack.optimize.CommonsChunkPlugin({
                    name: "_vendor",
                    minChunks: function(module) {
                        return /node_modules/.test(module.resource);
                    }
                }),

                new webpack.optimize.CommonsChunkPlugin({
                    name: "_manifest"
                }),

                new webpack.optimize.UglifyJsPlugin({
                    sourceMap: options.get("enabled.sourceMap")
                }),

                new OptimizeCssAssetsPlugin({
                    cssProcessor: require("cssnano"),
                    cssProcessorOptions: {
                        discardComments: { removeAll: true }
                    },
                    canPrint: true
                })
            ]
        });
    }
};
