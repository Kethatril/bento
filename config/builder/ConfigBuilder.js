const { Config } = require("webpack-config"),
    parse = require("./modules/parse");

module.exports = class ConfigBuilder {
    static create(options = {}) {
        return new ConfigBuilder(options);
    }

    constructor(options = {}) {
        this._options = parse(options);
        this._config = new Config();
        this._modules = [];
        this._manifest = {};
    }

    use(...modules) {
        this._modules = this._modules.concat(modules);
        return this;
    }

    build(fn) {
        this._modules.push(fn);

        return new Promise((resolve, reject) => {
            let i = -1,
                modules = this._modules,
                // Resolve array length to a valid (ToUint32) number.
                len = modules.length >>> 0;

            let config = new Config(),
                options = this._options;

            try {
                (function next(err) {
                    let async;

                    // Increment counter variable and skip any indices that
                    // don't exist. This allows sparse arrays to be iterated.
                    do {
                        ++i;
                    } while (!(i in modules) && i !== len);

                    if (err === false || err) return reject(err);

                    if (i === len) return resolve(config.toObject());

                    let result;

                    result = modules[i].call(
                        {
                            async() {
                                async = true;
                                return next;
                            }
                        },
                        config,
                        options
                    );
                    if (!async) {
                        next(result);
                    }
                })();
            } catch (e) {
                reject(e);
            }
        });
    }
};
