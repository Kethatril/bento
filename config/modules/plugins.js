const { config, read, format } = require("../utils"),
    path = require("path"),
    webpack = require("webpack");

const CleanPlugin = require("clean-webpack-plugin");
const CopyGlobsPlugin = require("copy-globs-webpack-plugin");
const WebpackAssetsManifest = require("webpack-assets-manifest");

module.exports = config(instance => {
    instance.merge({
        plugins: [
            new CleanPlugin([`${path.join(read("paths.output"), "**/*")}`], {
                root: read("root"),
                verbose: false
            }),

            new CopyGlobsPlugin({
                pattern: read("files.copy"),
                output: `[path]${read("filename")}.[ext]`,
                manifest: read("manifest")
            }),

            new WebpackAssetsManifest({
                output: "assets.json",
                space: 4,
                writeToDisk: false,
                sortManifest: true,
                assets: read("manifest"),
                replacer: format
            }),

            new webpack.DefinePlugin({
                "process.env": {
                    NODE_ENV: JSON.stringify(process.env.NODE_ENV)
                }
            })
        ]
    });

    if (read("env.isProduction")) {
        const OptimizeCssAssetsPlugin = require("optimize-css-assets-webpack-plugin");

        instance.merge({
            plugins: [
                new webpack.optimize.CommonsChunkPlugin({
                    name: "_vendor",
                    minChunks: function(module) {
                        return /node_modules/.test(module.resource);
                    }
                }),

                new webpack.optimize.CommonsChunkPlugin({
                    name: "_manifest"
                }),

                new webpack.optimize.UglifyJsPlugin({
                    sourceMap: read("enabled.sourceMap")
                }),

                new OptimizeCssAssetsPlugin({
                    cssProcessor: require("cssnano"),
                    cssProcessorOptions: {
                        discardComments: { removeAll: true }
                    },
                    canPrint: true
                })
            ]
        });
    }

    return instance;
});
