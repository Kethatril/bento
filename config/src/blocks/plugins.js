import webpack from "webpack";
import merge from "webpack-merge";
import CleanPlugin from "clean-webpack-plugin";
import CopyGlobsPlugin from "copy-globs-webpack-plugin";
import WebpackAssetsManifest from "webpack-assets-manifest";

export default function plugins(config, options, { manifest }) {
    return merge(config, {
        plugins: [
            new CleanPlugin([options.get("paths.output")], {
                root: options.get("paths.root"),
                verbose: false
            }),

            new CopyGlobsPlugin({
                pattern: options.get("files.copy"),
                output: `[path]${options.get("filename")}.[ext]`,
                manifest: manifest.data
            }),

            new WebpackAssetsManifest({
                output: "manifest.json",
                space: 4,
                writeToDisk: options.get("enabled.writeManifest"),
                sortManifest: true,
                assets: manifest.data,
                replacer: manifest.formatter(options.get("paths.public")),
                customize(key) {
                    return key.indexOf(".map") < 0;
                }
            }),

            new webpack.DefinePlugin({
                "process.env": {
                    NODE_ENV: JSON.stringify(options.get("env.value"))
                }
            }),

            new webpack.NamedModulesPlugin()
        ]
    });
}
