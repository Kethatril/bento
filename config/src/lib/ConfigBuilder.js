import { parse, resolve, select } from "~/lib/utils";
import ManifestKeeper from "~/lib/ManifestKeeper";

export default class ConfigBuilder {
    static create(options = {}) {
        return new ConfigBuilder(options);
    }

    manifest = new ManifestKeeper();

    constructor(options = {}) {
        this._options = parse(resolve(options));
        this._blocks = options.blocks.list || [];
        this._timeout = options.blocks.timeout || 3000;
    }

    get options() {
        return this._options;
    }

    build() {
        return new Promise((resolve, reject) => {
            let i = -1,
                modules = this._blocks,
                // Resolve array length to a valid (ToUint32) number.
                len = modules.length >>> 0;

            const timeout = this._timeout;

            let config = {},
                options = this._options,
                utils = { select: select(options), manifest: this.manifest };

            (function next(result) {
                let async;

                // Increment counter variable and skip any indices that
                // don't exist. This allows sparse arrays to be iterated.
                do {
                    ++i;
                } while (!(i in modules) && i !== len);

                if (result === false || result instanceof Error)
                    return reject(err);

                if (i === len) {
                    return resolve(result);
                }

                try {
                    result = modules[i](result, options, utils);
                } catch (e) {
                    return reject(e);
                }
                if (!result || typeof result !== "object") {
                    throw new Error(
                        "[InvalidValueError] Each block must either return a new config object or a promise that resolves a new config object."
                    );
                }

                if (result.then) {
                    const id = setTimeout(function() {
                        if (!async) return;
                        return reject(
                            new Error(
                                `[TimeoutError] ${timeout} ms exceeded. For async blocks, ensure the returned promise is resolved.`
                            )
                        );
                    }, timeout);

                    const done = function(result) {
                        clearTimeout(id);
                        next(result);
                    };

                    async = true;
                    result.then(done, done);
                }

                if (!async) {
                    next(result);
                }
            })(config);
        });
    }
}
