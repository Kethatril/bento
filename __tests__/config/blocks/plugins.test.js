import plugins from "~/blocks/plugins";
import ConfigBuilder from "~/lib/ConfigBuilder";

describe("blocks/plugins", () => {
    const options = {
        filename: "[name]",

        paths: {
            root: process.cwd(),
            input: "./app",
            output: "./public"
        },
        files: {
            copy: "+(images|media)/**/*"
        },
        blocks: {
            list: [plugins]
        },
        enabled: {
            writeManifest: true
        }
    };

    it("matches snapshot", () => {
        const builder = ConfigBuilder.create(Object.assign({}, options));

        return builder.build().then(config => {
            expect(config).toMatchSnapshot();
        });
    });
});
