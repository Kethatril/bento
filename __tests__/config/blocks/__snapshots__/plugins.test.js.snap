// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`blocks/plugins matches snapshot 1`] = `
Object {
  "plugins": Array [
    CleanWebpackPlugin {
      "options": Object {
        "allowExternal": false,
        "dry": false,
        "root": "<cwd>",
        "verbose": false,
      },
      "paths": Array [
        "<cwd>/public",
      ],
    },
     {
      "disable": undefined,
      "files": Array [],
      "globOptions": Object {
        "cwd": null,
        "nodir": true,
      },
      "manifest": Object {},
      "output": "[path][name].[ext]",
      "pattern": "+(images|media)/**/*",
      "watcher": null,
    },
    WebpackAssetsManifest {
      "_events": Object {},
      "_eventsCount": 0,
      "_maxListeners": undefined,
      "assets": Object {},
      "compiler": null,
      "domain": null,
      "hmrRegex": null,
      "options": Object {
        "contextRelativeKeys": false,
        "customize": null,
        "fileExtRegex": /\\.\\w{2,4}\\.(?:map|gz)$|\\.\\w+$/i,
        "merge": false,
        "output": "manifest.json",
        "publicPath": null,
        "replacer": [Function anonymous],
        "sortManifest": true,
        "space": 4,
        "writeToDisk": true,
      },
      "stats": null,
    },
    DefinePlugin {
      "definitions": Object {
        "process.env": Object {
          "NODE_ENV": undefined,
        },
      },
    },
    NamedModulesPlugin {
      "options": Object {},
    },
  ],
}
`;
